knapiks/mcchat.py

```diff
@@ -104,80 +106,79 @@ def process_current_log():
             continue
 
 
+# This absolutely must be renamed
 @app.task
 def check_for_players():
     # this is typically to be run every minute
     # check if anyone is logged in
     result = login_and_send('list')
-    players_without_logout_timestamps = []
-    all_players = Player.objects.all()
     player_pat = re.compile('[^ ]+')
 
+    process_current_log()
+
+    update_logouts(result)
+   
+    # process logins first and update the player model
+    update_active_players(unsent_logins)
+
+    # then grab chats and logouts
+    unsent_chats = Log.objects.filter(msg_content__startswith='<', msg_twilled=None)
+    logouts = Log.objects.filter(msg_content__endswith='left the game', msg_twilled=None)
+    msgs_to_send = unsent_chats | logouts
+
+    for msg in msgs_to_send:
+        pending_messages.append(msg.id)
+
+    send_messages()
+
+    print(result)
+
+def get_active_players:
     # collect list of players in local db who appear to be logged in based on login/logout timestamps
-    for player in all_players.exclude(last_logout=None):
-        if player.last_login > player.last_logout:
+    all_players = Player.objects.all()
+    players_without_logout_timestamps = []
+    for player in all_players:
+        if player.last_login > player.last_logout || player.last_logout == None:
             players_without_logout_timestamps.append(player)
 
-    for player in all_players.filter(last_logout=None):
-        players_without_logout_timestamps.append(player)
+    return players_without_logout_timestamps
+
 
+def update_active_players(unsent_logins):
+    unsent_logins = Log.objects.filter(msg_content__contains='joined the game', msg_twilled=None)
+    for msg in unsent_logins:
+        player_name = re.match(player_pat, msg.msg_content).group()
+        try:
+            player = Player.objects.get_or_create(name=player_name)[0]
+            player.last_login = msg.msg_time
+            player.save()
+
+            pending_messages.append(msg.id)
+        except Exception as e:
+            print(e)
+
+
+def update_logouts(result):
     # if no one is actually logged in, but the local db indicates someone may be logged in, process the current log
     # and mark them as logged out and send the sysop a message that they're logged out
+    players_without_logout_timestamps = get_active_players()
     if result == 'There are 0 of a max 20 players online: ' and len(players_without_logout_timestamps) > 0:
-        process_current_log()
         for player in players_without_logout_timestamps:
             last_logout = Log.objects.filter(msg_content=f'{player.name} left the game').last()
             player.last_logout = last_logout.msg_time
             player.save()
-            last_logout.msg_twilled = timezone.now()
-            last_logout.save()
-            sms = twilio_client.messages.create(
-                from_=f'+{twilio_phone}',
-                body=f'{player} logged out',
-                to=f'+{admin_phone}'
-            )
```

Comment:
I extracted this functionality and consolidated the sending of the messages to a single place

Reply:
Unless you absolutely require separate texts for logouts as well as the specific text "logged out" vs "left the game", my edits should simplify things